plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -94,22 +98,31 @@ public void executed(Data data) {
             }
 
             @Override
-            public void failed(Exception e) {
+            public void failed(Throwable e) {
                 logger.error("failed to execute pipeline [{}]", e, pipelineId);
                 listener.onFailure(e);
             }
         });
     }
 
-    void processBulkIndexRequest(String action, ActionListener listener, ActionFilterChain chain, BulkRequest bulkRequest, String pipelineId, Iterator<ActionRequest> requests) {
-        if (!requests.hasNext()) {
-            chain.proceed(action, bulkRequest, listener);
+    void processBulkIndexRequest(BulkRequestSource bulkRequestSource, String pipelineId, String action, ActionFilterChain chain, ActionListener<BulkResponse> listener) {
+        if (!bulkRequestSource.hasNext()) {
+            BulkRequest bulkRequest = bulkRequestSource.getBulkRequest();
+            ActionListener<BulkResponse> actionListener = bulkRequestSource.wrapActionListenerIfNeeded(listener);
+            if (bulkRequest.requests().isEmpty()) {
+                // in this stage, the transport bulk action can't deal with a bulk request with no requests,
+                // so we stop and send a empty response back to the client.
+                // (this will happen if all preprocessing all items in the bulk failed)
+                actionListener.onResponse(new BulkResponse(new BulkItemResponse[0], 0));
```

Comment:
with the current randomizations, do we ever test this branch? If not would it be worth to write a specific test for it?